### PR TITLE
Nudge students and mentors

### PR DESCRIPTION
**Pull request by ErikSchierboom**
Here is how to stub the notifications:

```ruby
student = User.for!('erikschierboom')
mentor = User.for!('ihid')

solution = student.solutions.published.last

req = Mentor::Request.create!(solution:, comment_markdown: "Could you please look at my code?")
discussion = Mentor::Discussion::Create.(mentor, req, 1, "Nice work!")
discussion.awaiting_mentor!
discussion.update(awaiting_mentor_since: Time.current - 8.days)
num_days_waiting = 7

User::Notification::Create.(
  discussion.mentor,
  :nudge_mentor_to_reply_in_discussion,
  discussion:,
  num_days_waiting:
)

discussion.awaiting_student!
discussion.update(awaiting_student_since: Time.current - 8.days)

User::Notification::Create.(
  discussion.mentor,
  :nudge_student_to_reply_in_discussion,
  discussion:,
  num_days_waiting:
)

User::Notification::Create.(
  discussion.mentor,
  :mentor_timed_out_discussion_student,
  discussion:
)

User::Notification::Create.(
  discussion.mentor,
  :mentor_timed_out_discussion_mentor,
  discussion:
)

User::Notification::Create.(
  discussion.mentor,
  :student_timed_out_discussion_student,
  discussion:
)

User::Notification::Create.(
  discussion.mentor,
  :student_timed_out_discussion_mentor,
  discussion:
)
```

### CODE REVIEW

File: app/models/user/notifications/nudge_mentor_to_reply_in_discussion_notification.rb [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+    {
+      student_name: student.handle,
+      track_title: track.title,
+      exercise_title: exercise.title
+    }
+  end
+
+  def image_type = :avatar
+  def image_url = student.avatar_url
+  def guard_params = "Discussion##{discussion.id}##{discussion.awaiting_mentor_since.to_date}##{num_days_waiting}"
```
Comment (iHiD) [inline app/models/user/notifications/nudge_mentor_to_reply_in_discussion_notification.rb:18]: Maybe?
```suggestion
  def guard_params = "Discussion##{discussion.id}##{discussion.awaiting_mentor_since.to_date.to_s}##{num_days_waiting}"
```

File: config/sidekiq-schedule.yml [hunk 2]
```diff
@@ -41,6 +41,11 @@ mark_outdated_reputation_year_periods:
   queue: reputation
   args: ["year"]
 
+send_mentor_discussion_nudges:
+  cron: "15 0 * * *"
```
Comment (ErikSchierboom) [inline config/sidekiq-schedule.yml]: @iHiD You wanted this to run every hour, correct?

Comment (iHiD) [reply config/sidekiq-schedule.yml]: Yes.

File: app/commands/mentor/discussion/send_nudges.rb [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+      Mentor::Discussion::NudgeStudent.(discussion, num_days_waiting)
+    rescue StandardError => e
+      Bugsnag.notify(e)
+    end
+  end
+
+  def student_nudge_discussions
+    Mentor::Discussion.includes(:student).
+      awaiting_student.
+      where('awaiting_student_since < ?', Time.now.utc - 7.days)
```
Comment (iHiD) [inline app/commands/mentor/discussion/send_nudges.rb:24]: We'll need an index on this.

File: app/mailers/notifications_mailer.rb [hunk 4]
```diff
@@ -221,4 +221,88 @@ def expired_insiders
...
+  def nudge_mentor_to_reply_in_discussion
+    notification = params[:notification]
+    @user = notification.user
+    @discussion = notification.discussion
+    @exercise = @discussion.exercise
+    @track = @exercise.track
+
+    @unsubscribe_key = :email_on_nudge_mentor_to_reply_in_discussion_notification
+
+    @title = "Your student is waiting for you to reply on your discussion"
```
Comment (iHiD) [inline app/mailers/notifications_mailer.rb]:
```suggestion
    @title = "Your student has been waiting #{notification. num_days_waiting} for your reply"
```

Comment (iHiD) [reply app/mailers/notifications_mailer.rb]: And the same for the other one.

File: app/views/mailers/notifications_mailer/nudge_mentor_to_reply_in_discussion.html.haml [hunk 5]
```diff
@@ -0,0 +1,5 @@
+%p Hi #{@user.handle},
+%p Your student (#{@discussion.student.handle}) is waiting for you to reply in the discussion on your solution to #{@exercise.title} on the #{@track.title} track.
```
Comment (iHiD) [inline app/views/mailers/notifications_mailer/nudge_mentor_to_reply_in_discussion.html.haml]:
```suggestion
%p This is a friendly reminder that your student (#{@discussion.student.handle}) has been waiting #{notification. num_days_waiting} days for you to reply in #{link_to "the discussion", track_exercise_mentor_discussion_url(@track, @exercise, @discussion)} on their solution to #{@exercise.title} on the #{@track.title} track.

%p If it's not your turn and you're waiting on the student, please use the "It's the student's turn…" button in the mentoring UI.
```

etc for the others.

File: app/commands/mentor/discussion/update_timed_out.rb [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+      where('awaiting_student_since < ?', timeout_date)
+  end
+
+  def update_mentor_timed_out!
+    mentor_timed_out_discussions.find_each do |discussion|
+      Mentor::Discussion::MentorTimeOut.(discussion)
+    rescue StandardError => e
+      Bugsnag.notify(e)
+    end
+  end
```
Comment (iHiD) [inline app/commands/mentor/discussion/update_timed_out.rb:30]: I have a preference for putting this method above the one above it as its referered to earlier in the file. I like keeping the action methods directly under `private` where possible.

File: app/mailers/notifications_mailer.rb [hunk 7]
```diff
@@ -221,4 +221,88 @@ def expired_insiders
...
+    notification = params[:notification]
+    @user = notification.user
+    @discussion = notification.discussion
+    @exercise = @discussion.exercise
+    @track = @exercise.track
+
+    @unsubscribe_key = :email_on_nudge_student_to_reply_in_discussion_notification
+
+    @title = "Your mentor has been waiting #{notification.num_days_waiting} days for your reply"
+    subject = "[Mentoring] #{@discussion.mentor.handle} has is waiting for you to reply to the discussion on #{@track.title}/#{@exercise.title}" # rubocop:disable Layout/LineLength
```
Comment (iHiD) [inline app/mailers/notifications_mailer.rb]:
```suggestion
    subject = "Your mentor has been waiting #{notification.num_days_waiting} days for your reply." # rubocop:disable Layout/LineLength
```

File: app/mailers/notifications_mailer.rb [hunk 8]
```diff
@@ -221,4 +221,88 @@ def expired_insiders
...
+    notification = params[:notification]
+    @user = notification.user
+    @discussion = notification.discussion
+    @exercise = @discussion.exercise
+    @track = @exercise.track
+
+    @unsubscribe_key = :email_on_nudge_mentor_to_reply_in_discussion_notification
+
+    @title = "Your student has been waiting #{notification.num_days_waiting} days for your reply"
+    subject = "[Mentoring] #{@discussion.student.handle} has is waiting for you to reply to the discussion on #{@track.title}/#{@exercise.title}" # rubocop:disable Layout/LineLength
```
Comment (iHiD) [inline app/mailers/notifications_mailer.rb]:
```suggestion
    subject = "[Mentoring] #{@discussion.student.handle} is waiting for you to reply (#{notification.num_days_waiting} day nudge)" # rubocop:disable Layout/LineLength
```

File: app/mailers/notifications_mailer.rb [hunk 9]
```diff
@@ -221,4 +221,88 @@ def expired_insiders
...
+    notification = params[:notification]
+    @user = notification.user
+    @discussion = notification.discussion
+    @exercise = @discussion.exercise
+    @track = @exercise.track
+
+    @unsubscribe_key = :email_on_mentor_timed_out_discussion_notification
+
+    @title = "Your discussion has timed out"
+    subject = "[Mentoring] the discussion on your solution to #{@track.title}/#{@exercise.title} has timed out"
```
Comment (iHiD) [inline app/mailers/notifications_mailer.rb]:
```suggestion
    subject = "The discussion on your solution to #{@track.title}/#{@exercise.title} has timed out"
```

File: app/mailers/notifications_mailer.rb [hunk 10]
```diff
@@ -221,4 +221,88 @@ def expired_insiders
...
+    notification = params[:notification]
+    @user = notification.user
+    @discussion = notification.discussion
+    @exercise = @discussion.exercise
+    @track = @exercise.track
+
+    @unsubscribe_key = :email_on_mentor_timed_out_discussion_notification
+
+    @title = "Your mentoring session has timed out"
+    subject = "[Mentoring] your mentoring session on the solution by the solution by #{@discussion.student.handle} to #{@track.title}/#{@exercise.title} has timed out" # rubocop:disable Layout/LineLength
```
Comment (iHiD) [inline app/mailers/notifications_mailer.rb]:
```suggestion
    subject = "[Mentoring] Your mentoring session has timed out due to lack of response by you." # rubocop:disable Layout/LineLength
```

File: app/mailers/notifications_mailer.rb [hunk 11]
```diff
@@ -221,4 +221,88 @@ def expired_insiders
...
+    notification = params[:notification]
+    @user = notification.user
+    @discussion = notification.discussion
+    @exercise = @discussion.exercise
+    @track = @exercise.track
+
+    @unsubscribe_key = :email_on_student_timed_out_discussion_notification
+
+    @title = "Your discussion has timed out"
+    subject = "[Mentoring] your discussion on #{@track.title}/#{@exercise.title} has timed out"
```
Comment (iHiD) [inline app/mailers/notifications_mailer.rb]:
```suggestion
    subject = "Your mentoring discussion on #{@track.title}/#{@exercise.title} has timed out"
```

File: app/mailers/notifications_mailer.rb [hunk 12]
```diff
@@ -221,4 +221,88 @@ def expired_insiders
...
+    notification = params[:notification]
+    @user = notification.user
+    @discussion = notification.discussion
+    @exercise = @discussion.exercise
+    @track = @exercise.track
+
+    @unsubscribe_key = :email_on_student_timed_out_discussion_notification
+
+    @title = "Your student's discussion has timed-out"
+    subject = "[Mentoring] your mentoring session on the solution by #{@discussion.student.handle} on #{@track.title}/#{@exercise.title} has timed out" # rubocop:disable Layout/LineLength
```
Comment (iHiD) [inline app/mailers/notifications_mailer.rb]:
```suggestion
    subject = "[Mentoring] Your mentoring session has timed out due to lack of student response." # rubocop:disable Layout/LineLength
```

File: app/views/mailers/notifications_mailer/nudge_mentor_to_reply_in_discussion.html.haml [hunk 13]
```diff
@@ -0,0 +1,8 @@
+%p Hi #{@user.handle},
+
+%p This is a friendly reminder that your student (#{@discussion.student.handle}) has been waiting #{notification.num_days_waiting} days for you to reply in #{link_to 'the discussion', track_exercise_mentor_discussion_url(@track, @exercise, @discussion)} on their solution to #{@exercise.title} on the #{@track.title} track.
```
Comment (iHiD) [inline app/views/mailers/notifications_mailer/nudge_mentor_to_reply_in_discussion.html.haml:3]:
```suggestion
%p This is a friendly reminder that your student (#{@discussion.student.handle}) has been waiting #{notification.num_days_waiting} days for you to reply in #{link_to 'the discussion', track_exercise_mentor_discussion_url(@track, @exercise, @discussion)} on their solution to #{@exercise.title} on the #{@track.title} track.
```

File: app/views/mailers/notifications_mailer/nudge_student_to_reply_in_discussion.html.haml [hunk 14]
```diff
@@ -0,0 +1,6 @@
+%p Hi #{@user.handle},
+
+%p This is a friendly reminder that your mentor (#{@discussion.student.handle}) has been waiting #{notification.num_days_waiting} days for you to reply in #{link_to 'the discussion', track_exercise_mentor_discussion_url(@track, @exercise, @discussion)} on your solution to #{@exercise.title} on the #{@track.title} track.
```
Comment (iHiD) [inline app/views/mailers/notifications_mailer/nudge_student_to_reply_in_discussion.html.haml]: Please check my logic here.

```suggestion
%p This is a friendly reminder that your mentor (#{@discussion.student.handle}) has been waiting #{notification.num_days_waiting} days for you to reply in #{link_to 'the discussion', track_exercise_mentor_discussion_url(@track, @exercise, @discussion)} on your solution to #{@exercise.title} on the #{@track.title} track.

If you are actually waiting for the mentor, please add a comment to the discussion to let them know.

The discussion will time out in #{28 - notification.num_days_waiting} days if you do not reply.
```

File: app/views/mailers/notifications_mailer/nudge_mentor_to_reply_in_discussion.text.haml [hunk 15]
```diff
@@ -0,0 +1,7 @@
+Hi #{@user.handle},
+\
+This is a friendly reminder that your student (#{@discussion.student.handle}) has been waiting #{notification.num_days_waiting} days for you to reply in the discussion on their solution to #{@exercise.title} on the #{@track.title} track.
+\
+If it's not your turn and you're waiting on the student, please use the "It's the student's turn…" button in the mentoring UI.
```
Comment (iHiD) [inline app/views/mailers/notifications_mailer/nudge_mentor_to_reply_in_discussion.text.haml:5]: Please check my logic here. Maybe we use a constant for 28?

```suggestion
If it's not your turn and you're waiting on the student, please use the "It's the student's turn…" button in the mentoring UI.

The discussion will time out in #{28 - notification.num_days_waiting} days if you do not reply.
```

File: db/migrate/20230830114417_add_email_on_nudge_student_to_reply_in_discussion_notification.rb [hunk 16]
```diff
@@ -0,0 +1,9 @@
+class AddEmailOnNudgeStudentToReplyInDiscussionNotification < ActiveRecord::Migration[7.0]
+  def change
+    return if Rails.env.production?
+
+    add_column :user_communication_preferences, :email_on_nudge_student_to_reply_in_discussion_notification, :boolean, default: true, null: false
+
+    # TODO: do we want to set this value to a default based on an existing preference?
```
Comment (iHiD) [inline db/migrate/20230830114417_add_email_on_nudge_student_to_reply_in_discussion_notification.rb]: No, default it to `true` for all of the below.

File: app/commands/mentor/discussion/send_nudges.rb [hunk 17]
```diff
@@ -0,0 +1,48 @@
...
+
+  def mentor_nudge_discussions
+    Mentor::Discussion.includes(:mentor).
+      awaiting_mentor.
+      where('awaiting_mentor_since < ?', Time.now.utc - 7.days)
+  end
+
+  def num_days_waiting_since(time)
+    NUM_DAYS_WAITING_CUTOFFS.find do |cutoff|
+      Time.now.utc.to_date - time.utc.to_date >= cutoff
```
Comment (iHiD) [inline app/commands/mentor/discussion/send_nudges.rb]: Only send within one day.

File: app/models/mentor/discussion.rb [hunk 18]
```diff
@@ -162,6 +165,40 @@ def awaiting_mentor!
...
+    update!(cols)
+  end
+
+  def student_abandoned!
+    update!(
+      status: :finished,
+      awaiting_mentor_since: nil,
+      awaiting_student_since: nil
+    )
+  end
```
Comment (iHiD) [inline app/models/mentor/discussion.rb]: I'd rather we moved all of these into the commands. I think we never want these to just arbitrarily be called without being in the commands. Let me know if you disagree and we can discuss.

Comment (ErikSchierboom) [reply app/models/mentor/discussion.rb]: Sure! I just followed what was already there.

File: app/commands/mentor/discussion/finish_by_mentor.rb [hunk 19]
```diff
@@ -4,11 +4,23 @@ class Mentor::Discussion::FinishByMentor
...
   end
 
   private
+  def update!
+    discussion.update!(
+      status: :mentor_finished,
+      finished_at: Time.current,
+      finished_by: :mentor,
+      awaiting_mentor_since: nil,
+      awaiting_student_since: discussion.awaiting_student_since || Time.current
```
Comment (iHiD) [inline app/commands/mentor/discussion/finish_by_mentor.rb:23]: ~I think this should be `nil`.~

Actually, I quite like what you've done here on second thoughts.

File: app/commands/mentor/discussion/mentor_timed_out.rb [hunk 20]
```diff
@@ -0,0 +1,38 @@
+class Mentor::Discussion::MentorTimedOut
+  include Mandate
+
+  initialize_with :discussion
+
+  def call
+    update!
+    create_notifications!
```
Comment (iHiD) [inline app/commands/mentor/discussion/mentor_timed_out.rb]: Can you wrap this in a transaction pls?

File: app/commands/mentor/discussion/finish_by_student.rb [hunk 21]
```diff
@@ -36,6 +34,20 @@ def call
   end
```
Comment (iHiD) [inline app/commands/mentor/discussion/finish_by_student.rb:37]: Can you wrap the above in a transaction pls?

File: app/commands/mentor/discussion/finish_by_mentor.rb [hunk 22]
```diff
@@ -4,11 +4,23 @@ class Mentor::Discussion::FinishByMentor
   initialize_with :discussion
 
   def call
-    discussion.mentor_finished!
+    return if discussion.finished?
+
+    update!
     notify!
```
Comment (iHiD) [inline app/commands/mentor/discussion/finish_by_mentor.rb:13]: Can you wrap these in a transaction pls?

File: app/controllers/tracks/mentor_discussions_controller.rb [hunk 23]
```diff
@@ -6,7 +6,8 @@ class Tracks::MentorDiscussionsController < ApplicationController
   def index
     return redirect_to track_exercise_path(@track, @exercise) unless @solution&.iterated?
 
-    @previous_discussions = @solution.mentor_discussions.finished.includes(:mentor)
+    @previous_discussions = @solution.mentor_discussions.includes(:mentor).where(status: %i[finished student_timed_out
+                                                                                            mentor_timed_out])
```
Comment (iHiD) [inline app/controllers/tracks/mentor_discussions_controller.rb]:
```suggestion
    @previous_discussions = @solution.mentor_discussions.includes(:mentor).
      where(status: %i[finished student_timed_out mentor_timed_out])
```

File: app/jobs/finish_abandoned_job.rb [hunk 24]
```diff
@@ -0,0 +1,7 @@
+# The goal of this job is to re-sync sponsors to ensure
```
Comment (iHiD) [inline app/jobs/finish_abandoned_job.rb]: This class should probably be renamed to `FinishAbandonedMentorDiscussionsJob` (add "Mentor") and the file needs to be named to match.

File: app/jobs/send_discussion_nudges_job.rb [hunk 25]
```diff
@@ -0,0 +1,5 @@
+class SendDiscussionNudgesJob < ApplicationJob
```
Comment (iHiD) [inline app/jobs/send_discussion_nudges_job.rb]:
```suggestion
class SendMentorDiscussionNudgesJob < ApplicationJob
```

File: yarn.lock [hunk 26]
```diff
@@ -9295,11 +9292,6 @@ source-map@^0.7.3:
   resolved "https://registry.yarnpkg.com/source-map/-/source-map-0.7.3.tgz#5302f8169031735226544092e64981f751750383"
   integrity sha512-CkCj6giN3S+n9qrYiBTX5gystlENnRW5jZeNLHpe6aue+SrHcG5VYwujhW9s4dY31mEGsxBDrHR6oI69fTXsaQ==
 
-spark-md5@^3.0.0:
-  version "3.0.2"
-  resolved "https://registry.yarnpkg.com/spark-md5/-/spark-md5-3.0.2.tgz#7952c4a30784347abcee73268e473b9c0167e3fc"
-  integrity sha512-wcFzz9cDfbuqe0FZzfi2or1sgyIrsDwmPwfZC4hiNidPdPINjeUwNfv5kldczoEAcjl9Y1L3SM7Uz2PUEQzxQw==
-
```
Comment (iHiD) [inline yarn.lock]: You might want to revert all these `yarn.lock` changes on this PR.

**Pull request merged**